reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java

```diff
@@ -91,17 +91,17 @@ public void subscribe(Subscriber<? super T> s) {
 		Objects.requireNonNull(s, "s");
 
 		ColdTestPublisherSubscription<T> p = new ColdTestPublisherSubscription<>(s, this);
-		s.onSubscribe(p); // will trigger drain() via request()
-
 		ColdTestPublisher.SUBSCRIBED_COUNT.incrementAndGet(this);
 
-		if (p.cancelled) {
-			return;
-		}
-
 		if (!add(p)) {
-			p.drain();
+			s.onSubscribe(p); // will trigger drain() via request()
```

Comment:
why `onSubscribe` at all if the subscriber cannot be added?

Reply:
because this is a cold subscriber, so in any case, it should be notified with all the values. The subscriber can not be added when TestPublisher is terminated with onComplete or onError